Improve Extension Documentation (#19727)
directus/directus: docs/extensions/installing-extensions.md

---
contributors: Esther Agbaje
description: Discover how to install extensions into your Directus project.
---

# Installing Extensions

There are two possible ways to install extensions to your Directus instance:

1. Installing through NPM
2. Installing through the Extensions Folder

> ## Installing through NPM

Reviewer: ```suggestion
## Installing Through NPM
```

If you wish to install an existing public extension, package extensions are essentially npm packages that can be
installed from the npm registry. Before you begin, ensure you have a
[self-hosted instance of Directus](/self-hosted/quickstart) via Docker installed on your system.

**1. Modify docker-compose.yml**

Open the `docker-compose.yml` file of your project and modify the build section:

```yaml
build:
  context: ./
```

This allows Docker to build your project with the necessary dependencies.

**2. Create a Dockerfile**

At the root of your project, create a `Dockerfile` if one doesn't already exist and add the following:

```Dockerfile
FROM directus/directus:latest

USER root
RUN corepack enable \
 && corepack prepare pnpm@8.7.6 --activate

USER node
RUN pnpm install directus-extension-package-name
```

::: tip Extension Name

Remember to replace `directus-extension-package-name` with the name of the extension you want to install. For example,
`directus-extension-myextension`.

:::

**3. Build the Docker Image**

Build your Docker image:

```bash

> docker compose --build

Reviewer: ```suggestion
docker compose build
```

```

**4. Start the Docker Container**

Start your Docker container:

```bash
docker compose up
```

On startup, you'd see that Directus will automatically load any package extension installed into your Directus project
folder.

## Installing Through the Extensions Folder

> To locally install extensions, copy the files generated by the `directus-extension build` command into the 'extensions'

Reviewer: ```suggestion
To locally install extensions, copy the files generated by the `directus-extension build` command into the "extensions"
```

---
contributors: Esther Agbaje
description: Discover how to install extensions into your Directus project.
---

# Installing Extensions

There are two possible ways to install extensions to your Directus instance:

1. Installing through NPM
2. Installing through the Extensions Folder

> ## Installing through NPM

Reviewer: Should we add a note here that this installation method makes sense if you want to install an existing public extension or if have published (or plan to do so) your own extension on npm?

Package extensions are essentially npm packages that can be installed from the npm registry. Before you begin, ensure
you have a [self-hosted instance of Directus](/self-hosted/quickstart) via Docker installed on your system.

**1. Modify docker-compose.yml**

Open the `docker-compose.yml` file of your project and modify the build section:

```yaml
build:
  context: ./
```

This allows Docker to build your project with the necessary dependencies.

**2. Create a Dockerfile**

At the root of your project, create a `Dockerfile` if one doesn't already exist and add the following:

```Dockerfile
FROM directus/directus:latest

USER root
RUN corepack enable \

>  && corepack prepare pnpm@8.3.1 --activate

Reviewer: ```suggestion
 && corepack prepare pnpm@8.7.6 --activate
```

USER node
RUN pnpm install directus-extension-package-name
```

::: tip Extension Name

Remember to replace `directus-extension-package-name` with the name of the extension you want to install. For example,
`directus-extension-myextension`.

:::

**3. Build the Docker Image**

Build your Docker image:

```bash
docker compose up --build
```

**4. Start the Docker Container**

Start your Docker container:

```bash
docker compose up

> ```

Reviewer: `docker compose up --build` already starts the container so these commands should either be combined or otherwise the first one should be replaced with `docker compose build`.

On startup, you'd see that Directus will automatically load any package extension installed into your Directus project
folder.

::: tip Automatically Reload Extensions

To automatically reload extensions every time you make a change, without having to restart Directus, in your
`docker-compose.yml` file, set `EXTENSIONS_AUTO_RELOAD=true`.

> :::

Reviewer: Should be moved to the "Installing Through the Extensions Folder" as it makes more sense there

## Installing Through the Extensions Folder

Local extensions are essentially the files generated by the `directus-extension build` command. They can be installed by

> copying those files into the extensions folder at the root of your Directus project.

Reviewer: Needs to be revised but am not sure myself how best to phrase it:
- An extension essentially consists of the folder and its files created by `create-directus-extension` 
- But this is also true for "Package extensions" (above) so that's not what makes a local extension
- it is actually only the form of installation that differs. That is, how the extension is made available to Directus -> Put extension into local folder vs. publish it to npm and install it from there

::: warning Configurable Folders

The path to the built extension as well as the extensions directory are configurable and may be located elsewhere.

:::

Before you begin, ensure you have a [self-hosted instance of Directus](/self-hosted/quickstart) via Docker installed on
your system.

**1. Create an Extension Folder**

At the root of your project, create an "extensions" folder if one doesn't already exist to house your extension.

**2. Add your new extension into the extensions folder**

Move the output from the `dist/` folder into the extensions folder you created earlier.

Your folder structure should look like this:

```yaml

> extensions
>   <extension-name>/
>     dist/
>       index.js
>     src
>       index.js
>       interface.vue
>   displays/
>   hooks/
>   endpoints/

Reviewer: The `package.json` needs to present too, not so the `src` directory.
But I'm not sure whether we should actively advise users to only copy those required files or simply let them copy the whole extension into the `extensions` folder.

```suggestion
extensions
  <extension-name>/
    dist/
      index.js
    package.json
```

> Move the output from the `dist/` folder into the extensions folder you created earlier.
> 
> Your folder structure should look like this:
> 
> ```yaml
> extensions
>   <extension-name>/
>     dist/
>       index.js
>     src
>       index.js
>       interface.vue
>   displays/
>   hooks/
>   endpoints/

Reviewer: The `package.json` needs to present too, not so the `src` directory.
But I'm not sure whether we should actively advise users to only copy those required files or simply let them copy the whole extension into the `extensions` folder 🤔 
The former is purely an optional optimization step.

```suggestion
Move the `package.json` file along with the output from the `dist/` folder into a new folder in the extensions folder you created earlier.

Your folder structure should look like this:

```yaml
extensions/
  <extension-name>/
    dist/
      index.js
    package.json
```

  ...
```

::: tip Extension Name Format

Ensure your extension is prefixed with `directus-extension`. For example, `directus-extension-myextension`.

:::

**3. Update Docker Compose File**

> Open your `docker-compose.yml` file and add the following to mount your extension into the Docker container:

Reviewer: ```suggestion
Open your `docker-compose.yml` file and add a volume to mount your extensions folder into the Docker container:
```

---
contributors: Esther Agbaje
description: Discover how to install extensions into your Directus project.
---

# Installing Extensions

There are three possible ways to install extensions to your Directus instance:

1. Installing through NPM
2. Installing through the Extensions Folder
3. Installing through the Marketplace

## Installing through NPM

Package extensions are essentially npm packages that can be installed from the npm registry. Before you begin, ensure
you have a [self-hosted instance of Directus](/self-hosted/quickstart) via Docker installed on your system.

> **1. Modify docker-compose.yml**

Reviewer 2: This needs a bit more context with perhaps some more steps to make work.

Author: I don't get what you mean here

Reviewer 2: Where does this modification go?

Author: This modification happens in the `docker-compose.yml` file under the build command. Is that what you mean?

Open the `docker-compose.yml` file of your project and modify the build section:

```yaml
build:
  context: ./
```

This allows Docker to build your project with the necessary dependencies.

**2. Create a Dockerfile**

At the root of your project, create a `Dockerfile` if one doesn't already exist and add the following:

```Dockerfile
FROM directus/directus:latest

USER root
RUN corepack enable \
 && corepack prepare pnpm@8.3.1 --activate

USER node
RUN pnpm install directus-extension-package-name
```

::: tip Extension Name

Remember to replace `directus-extension-package-name` with the name of the extension you want to install. For example,
`directus-extension-myextension`.

:::

**3. Build the Docker Image**

Build your Docker image:

```bash
docker-compose up --build
```

**4. Start the Docker Container**

Start your Docker container:

```bash

> docker-compose up

Reviewer 2: ```suggestion
docker compose up
```

copying those files into the extensions folder at the root of your Directus project.

::: warning Configurable Folders

The path to the built extension as well as the extensions directory are configurable and may be located elsewhere.

:::

Before you begin, ensure you have a [self-hosted instance of Directus](/self-hosted/quickstart) via Docker installed on
your system.

**1. Create an Extension Folder**

At the root of your project, create an "extensions" folder if one doesn't already exist to house your extension.

**2. Add your new extension into the extensions folder**

Move the output from the `dist/` folder into the extensions folder you created earlier.

Your folder structure should look like this:

```yaml
extensions
  <extension-name>/
    dist/
      index.js
    src
      index.js
      interface.vue
  displays/
  hooks/
  endpoints/
  ...
```

::: tip Extension Name Format

Ensure your extension is prefixed with `directus-extension`. For example, `directus-extension-myextension`.

:::

**3. Update Docker Compose File**

Open your `docker-compose.yml` file and add the following to mount your extension into the Docker container:

```yaml
volumes:

> 

Reviewer 2: Extra line

  - ./extensions:/directus/extensions/
```

**4. Start the Docker Container**

Start your Docker container:

```bash
docker-compose up
```

You should see that your extension has been successfully loaded into the Docker container. Now, go ahead to customize
your extension by making changes to the `src` folder within your extension directory.

## Installing through the Marketplace

Within your Directus Cloud account, you can conveniently install extensions from the Directus Marketplace.

**1. Access the Marketplace**

Log in to your [Directus Cloud account](https://directus.cloud). Once logged in, navigate to the "Marketplace" menu from
the Module Bar.

**2. Install Extensions**

Within the Directus Marketplace, you'll notice a list of extensions that can be integrated into your account. Search for
an extension that suits your needs and install it.

::: tip Admin Access

Only Administrators of a Directus instance can install extensions.

> :::

Reviewer 2: Remove this section
